aten/src/ATen/AccumulateType.h

```diff
@@ -51,147 +52,94 @@
-};
-template <>
-struct AccumulateType<Float8_e4m3fn, false> {
-  using type = float;
-};
-template <>
-struct AccumulateType<c10::complex<Half>, false> {
-  using type = c10::complex<float>;
+template <typename T>
+struct AccumulateType<T, false> {
+  using type = typename AccumulateTypeDevice<T, c10::DeviceType::CPU>::type;
 };
-template <>
-struct AccumulateType<c10::complex<float>, false> {
-  using type = c10::complex<double>;
-};
-template <>
-struct AccumulateType<c10::complex<double>, false> {
-  using type = c10::complex<double>;
-};
-template <>
-struct AccumulateType<c10::complex<Half>, true> {
-  using type = c10::complex<float>;
-};
-template <>
-struct AccumulateType<c10::complex<float>, true> {
-  using type = c10::complex<float>;
-};
-template <>
-struct AccumulateType<c10::complex<double>, true> {
-  using type = c10::complex<double>;
-};
-template <>
-struct AccumulateType<float, false> {
-  using type = double;
-};
-template <>
-struct AccumulateType<double, false> {
-  using type = double;
-};
-template <>
-struct AccumulateType<int8_t, false> {
-  using type = int64_t;
-};
-template <>
-struct AccumulateType<uint8_t, false> {
-  using type = int64_t;
-};
-template <>
-struct AccumulateType<char, false> {
-  using type = int64_t;
-};
-template <>
-struct AccumulateType<int16_t, false> {
-  using type = int64_t;
-};
-template <>
-struct AccumulateType<int32_t, false> {
-  using type = int64_t;
-};
-template <>
-struct AccumulateType<int64_t, false> {
-  using type = int64_t;
-};
-template <>
-struct AccumulateType<bool, false> {
-  using type = bool;
+
+template <typename T>
+struct AccumulateType<T, true> {
+  using type = typename AccumulateTypeDevice<T, c10::DeviceType::CUDA>::type;
 };
 
+template <typename T, c10::DeviceType device>
+using acc_type_device = typename AccumulateTypeDevice<T, device>::type;
+
 template <typename T, bool is_cuda>
 using acc_type = typename AccumulateType<T, is_cuda>::type;
 
+#define ACC_TYPE(t, acc_t, device_type)         \
+  template <>                                   \
+  struct AccumulateTypeDevice<t, device_type> { \
+    using type = acc_t;                         \
+  };
+#define MPS_ACC_TYPE(t, acc_t) ACC_TYPE(t, acc_t, c10::DeviceType::MPS)
+#define CUDA_ACC_TYPE(t, acc_t) ACC_TYPE(t, acc_t, c10::DeviceType::CUDA)
+#define CPU_ACC_TYPE(t, acc_t) ACC_TYPE(t, acc_t, c10::DeviceType::CPU)
+
+MPS_ACC_TYPE(BFloat16, float);
+MPS_ACC_TYPE(Half, float);
+MPS_ACC_TYPE(Float8_e5m2, float);
+MPS_ACC_TYPE(Float8_e4m3fn, float);
+MPS_ACC_TYPE(float, float);
+MPS_ACC_TYPE(double, float);
+MPS_ACC_TYPE(int8_t, int64_t);
+MPS_ACC_TYPE(uint8_t, int64_t);
+MPS_ACC_TYPE(char, int64_t);
+MPS_ACC_TYPE(int16_t, int64_t);
+MPS_ACC_TYPE(int32_t, int64_t);
+MPS_ACC_TYPE(int64_t, int64_t);
+MPS_ACC_TYPE(bool, bool);
+MPS_ACC_TYPE(c10::complex<Half>, c10::complex<float>);
+MPS_ACC_TYPE(c10::complex<float>, c10::complex<float>);
+MPS_ACC_TYPE(c10::complex<double>, c10::complex<float>);
```

Comment:
cc @kulinseth just to confirm that this is ok to have this for now.

Reply:
Can we issue a warning that we are falling back to using float for double types ?

Reply:
I'm a bit concerned about that would be somewhat verbose from a user's point of view since accumulate type is an internal thing of which a regular user probably doesn't have idea.

What do you think @albanD?

Reply:
I'm not sure what is the best place to put this warning for sure. Also note that the user would need to have created a compleex<double> to begin with to reach this.
So I would say that not having a warning here is fine?

Reply:
If the current state is fine, this PR needs your approval @albanD. Thanks!